<a href="https://colab.research.google.com/github/gnjohnson29/MPM_200_Johnson/blob/main/Session%209/assignment_9_database_management.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Python Packages and Scientific Computing Libraries

"""
Python Packages Overview:
-------------------------
Packages in Python are collections of modules that provide reusable code and functionality.
They allow developers to organize and share complex code libraries that can be easily imported
and used across different projects.

Key Characteristics of Packages:
- Organized hierarchically
- Contain multiple related modules
- Can be installed using package managers like pip
- Extend Python's core functionality

Three Essential Data Science Libraries:
--------------------------------------
1. Pandas (pd):
   - Primary library for data manipulation and analysis
   - Provides DataFrame: a 2D labeled data structure with columns of potentially different types
   - Key Features:
     * Reading/writing data from various sources (CSV, Excel, SQL databases)
     * Data cleaning and preprocessing
     * Complex data transformations
     * Powerful data analysis and aggregation tools

2. NumPy (np):
   - Fundamental package for scientific computing in Python
   - Introduces the ndarray (n-dimensional array) for efficient numerical operations
   - Key Features:
     * High-performance multidimensional array object
     * Mathematical and statistical functions
     * Linear algebra operations
     * Foundation for scientific and numerical computing
     * Enables fast computations on large datasets

3. Matplotlib (plt):
   - Comprehensive library for creating static, animated, and interactive visualizations
   - Provides MATLAB-like plotting interface
   - Key Features:
     * Line plots, scatter plots, bar charts, histograms
     * Customizable plot styles and layouts
     * Support for multiple plot types and graph formats
     * Integration with NumPy for data visualization

Practical Significance:
----------------------
These libraries form the backbone of data science and scientific computing in Python,
enabling complex data analysis, numerical computations, and insightful visualizations
with minimal code.
"""

In [1]:
# Import required libraries
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

# Read the Excel file
df = pd.read_excel("/content/Flavivirus host associations.xlsx")

# Display basic information about the dataset
print(df.info())

# Show the first few rows of the data
df.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 570 entries, 0 to 569
Data columns (total 10 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   sr no.                  570 non-null    int64  
 1   Virus                   570 non-null    object 
 2   Scientific Name         570 non-null    object 
 3   Authors                 570 non-null    object 
 4   Year                    430 non-null    object 
 5   Title/accession number  570 non-null    object 
 6   Journal                 425 non-null    object 
 7   Issue                   415 non-null    float64
 8   Volume                  404 non-null    object 
 9   Pages                   404 non-null    object 
dtypes: float64(1), int64(1), object(8)
memory usage: 44.7+ KB
None


sr no.         Virus     Scientific Name  \
0       3    Apoi virus       Rattus rattus   
1       2    Apoi virus    Myodes rufocanus   
2       0    Apoi virus  Apodemus argenteus   
3       1    Apoi virus  Apodemus speciosus   
4       4  Bagaza virus      Alectoris rufa   

                                             Authors  Year  \
0  C. M. Baak-Baak, N. Cigarroa-Toledo, O. M. Tor...  2016   
1                                        CDC Arbocat   NaN   
2                 J. S. Mackenzie and D. T. Williams  2009   
3                 J. S. Mackenzie and D. T. Williams  2009   
4  D. Buitrago, A. Rocha, C. Tena-Tomas, M. Vigo,...  2012   

                              Title/accession number  \
0  New Records of Ectoparasites Echinolaelaps ech...   
1                                        CDC Arbocat   
2  The zoonotic flaviviruses of southern, south-e...   
3  The zoonotic flaviviruses of southern, south-e...   
4  Real-time fluorogenic reverse transcription po...   

                     Journal  Issue Volume   Pages  
0  Southwestern Entomologist    NaN    NaN     NaN  
1                        NaN    NaN    NaN     NaN  
2     Zoonoses Public Health   56.0    6-7  338-56  
3     Zoonoses Public Health   56.0    6-7  338-56  
4         J Vet Diagn Invest   24.0      5  959-63

## Using pandas functions that you have seen in the class address following questions
1. How many entries are there in the dataset?
**570**
2. How many viruses are present in the dataset?
**30**
3. List those viruses.
**See answer with code below**
4. How many rows of data are present for each virus?
**See answer with code below**
5. How many host species are present in the data?
**419**
6. Make a list of host species present.
**See answer with code below**
7. Count the number of unique host species for each virus.
**See answer with code below**
8. Which virus has the most number of hosts?
**West Nile Virus**
9. Which host has most number of viruses?
**Equus Caballus**

In [2]:
df.columns.tolist()

['sr no.',
 'Virus',
 'Scientific Name',
 'Authors',
 'Year',
 'Title/accession number',
 'Journal',
 'Issue',
 'Volume',
 'Pages']

In [3]:
df.dtypes

sr no.                      int64
Virus                      object
Scientific Name            object
Authors                    object
Year                       object
Title/accession number     object
Journal                    object
Issue                     float64
Volume                     object
Pages                      object
dtype: object

In [4]:
#Answer to Q1"
total_entries = len(df)
print(f"Total number of entries in the dataset: {total_entries}")

Total number of entries in the dataset: 570


In [5]:
#Answer to Q2"
unique_viruses_count = df['Virus'].nunique()
print(f"Number of unique viruses in the dataset: {unique_viruses_count}")

Number of unique viruses in the dataset: 30


In [6]:
#Answer to Q3"
unique_viruses_names = df['Virus'].unique()
print("\nList of unique virus names:")
for virus_name in unique_viruses_names:
    print(f"-{virus_name}")


List of unique virus names:
-Apoi virus
-Bagaza virus
-Banzi virus
-Bouboui virus
-Bussuquara virus
-Cacipacore virus
-Dakar bat virus
-Dengue virus
-Entebbe bat virus
-Iguape virus
-Ilheus virus
-Japanese encephalitis virus
-Koutango virus
-Kunjin virus
-Kyasanur forest virus
-Louping-ill virus
-Modoc virus
-Murray Valley encephalitis virus
-Omsk hemorrhagic fever virus
-Rio Bravo virus
-Rocio virus
-St. Louis Encephalitis virus
-Tembusu virus
-Tickborne encephalitis virus
-Uganda S virus
-Usutu virus
-Wesselsbron virus
-West Nile virus
-Yellow fever virus
-Zika virus


In [7]:
#Answer to Q4"
virus_counts = df['Virus'].value_counts()
print("\nOccurrences of each unique virus:")
print(virus_counts)


Occurrences of each unique virus:
Virus
West Nile virus                     194
Tickborne encephalitis virus         84
St. Louis Encephalitis virus         72
Usutu virus                          65
Dengue virus                         27
Japanese encephalitis virus          25
Louping-ill virus                    13
Yellow fever virus                   11
Rio Bravo virus                       9
Zika virus                            7
Uganda S virus                        7
Entebbe bat virus                     6
Kyasanur forest virus                 6
Dakar bat virus                       5
Ilheus virus                          5
Apoi virus                            4
Modoc virus                           3
Murray Valley encephalitis virus      3
Wesselsbron virus                     3
Omsk hemorrhagic fever virus          3
Kunjin virus                          3
Bussuquara virus                      3
Bagaza virus                          2
Banzi virus                           2

In [8]:
#Answer to Q5"
unique_host_species_count = df['Scientific Name'].nunique()
print(f"Total number of unique host species: {unique_host_species_count}")

Total number of unique host species: 419


In [9]:
#Answer to Q6"
unique_host_species_names = df['Scientific Name'].unique()
print("\nList of unique host species names:")
for host_species_name in unique_host_species_names:
    print(f"-{host_species_name}")


List of unique host species names:
-Rattus rattus
-Myodes rufocanus
-Apodemus argenteus
-Apodemus speciosus
-Alectoris rufa
-Phasianus colchicus
-Mastomys natalensis
-Mesocricetus auratus
-Papio papio
-Equus caballus
-Alouatta belzebul
-Proechimys guyannensis
-Chaerephon pumilus
-Taphozous perforatus
-Mops condylurus
-Scotophilus nigrita
-Otomops sp
-Carollia brevicauda
-Molossus molossus
-Myotis nigricans
-Natalus stramineus
-Rousettus leschenaultii
-Artibeus planirostris
-Caluromys philander
-Carollia perspicillata
-Didelphis marsupialis
-Hylaeamys megacephalus
-Marmosa demerarae
-Marmosa murina
-Marmosops parvidens
-Mesomys hispidus
-Oecomys sp
-Philander opossum
-Proechimys cuvieri
-Zygodontomys brevicauda
-Artibeus jamaicensis
-Artibeus lituratus
-Glossophaga soricina
-Artibeus sp
-Uroderma sp
-Macaca nemestrina
-Mus musculus
-Eidolon helvum
-Pipistrellus pipistrellus
-Rhinolophus sp
-Rousettus spinalatus
-Macaca mulatta
-Equus ferus
-Molothrus bonariensis
-Sporophila caerulescen

In [10]:
#Answer to Q7"
virus_host_diversity = df.groupby('Virus')['Scientific Name'].nunique().sort_values(ascending=False)
print("Diversity of host species for each virus:")
print(virus_host_diversity)

Diversity of host species for each virus:
Virus
West Nile virus                     194
Tickborne encephalitis virus         83
St. Louis Encephalitis virus         72
Usutu virus                          65
Dengue virus                         27
Japanese encephalitis virus          25
Louping-ill virus                    13
Yellow fever virus                   11
Rio Bravo virus                       9
Zika virus                            7
Entebbe bat virus                     6
Kyasanur forest virus                 6
Ilheus virus                          5
Dakar bat virus                       5
Apoi virus                            4
Uganda S virus                        4
Modoc virus                           3
Murray Valley encephalitis virus      3
Wesselsbron virus                     3
Omsk hemorrhagic fever virus          3
Kunjin virus                          3
Bussuquara virus                      3
Bagaza virus                          2
Banzi virus                     

In [11]:
#Answer to Q8"
most_diverse_virus = virus_host_diversity.index[0]
num_unique_hosts = virus_host_diversity.iloc[0]
print(f"\nVirus with the most diverse host species: {most_diverse_virus} ({num_unique_hosts} unique hosts)")


Virus with the most diverse host species: West Nile virus (194 unique hosts)


In [12]:
#Answer to Q9"
most_common_host = df['Scientific Name'].value_counts().idxmax()
print(f"\nHost with the most common virus: {most_common_host}")


Host with the most common virus: Equus caballus


In [ ]:
# Exploring Virus and Species Diversity in the Dataset

# Count total number of entries in the dataset

# Count unique viruses

# List of unique viruses

# Count of each unique virus

# Count unique host species

# List of unique host species

# Count of host species for each virus

# Most diverse viruses (viruses with most unique host species)


Total number of entries in the dataset: 570
Number of unique viruses: 30

List of Unique Viruses:
['Apoi virus' 'Bagaza virus' 'Banzi virus' 'Bouboui virus'
 'Bussuquara virus' 'Cacipacore virus' 'Dakar bat virus' 'Dengue virus'
 'Entebbe bat virus' 'Iguape virus' 'Ilheus virus'
 'Japanese encephalitis virus' 'Koutango virus' 'Kunjin virus'
 'Kyasanur forest virus' 'Louping-ill virus' 'Modoc virus'
 'Murray Valley encephalitis virus' 'Omsk hemorrhagic fever virus'
 'Rio Bravo virus' 'Rocio virus' 'St. Louis Encephalitis virus'
 'Tembusu virus' 'Tickborne encephalitis virus' 'Uganda S virus'
 'Usutu virus' 'Wesselsbron virus' 'West Nile virus' 'Yellow fever virus'
 'Zika virus']

Virus Occurrence Counts:
Virus
West Nile virus                     194
Tickborne encephalitis virus         84
St. Louis Encephalitis virus         72
Usutu virus                          65
Dengue virus                         27
Japanese encephalitis virus          25
Louping-ill virus                    13
Y

# Metadata

## Given that we are interested in the metadata associated with these host species, we will import datasets associated with the metadata for both bird and mammalian species.

1. Birds

In [13]:
loc = "/content/Handbook of the Birds of the World and BirdLife International Digital Checklist of the Birds of the World_Version_9.xlsx"
birds_meta = pd.read_excel(loc)
birds_meta.head()

Seq.  Subsp. Seq.             Order    Family name     Family Subfamily  \
0   1.0          0.0  STRUTHIONIFORMES  Struthionidae  Ostriches       NaN   
1   1.0          1.0               NaN            NaN        NaN       NaN   
2   1.0          2.0               NaN            NaN        NaN       NaN   
3   1.0          3.0               NaN            NaN        NaN       NaN   
4   1.0          4.0               NaN            NaN        NaN       NaN   

  Tribe     Common name             Scientific name       Authority  \
0   NaN  Common Ostrich            Struthio camelus  Linnaeus, 1758   
1   NaN             NaN   Struthio camelus syriacus             NaN   
2   NaN             NaN    Struthio camelus camelus             NaN   
3   NaN             NaN  Struthio camelus massaicus             NaN   
4   NaN             NaN  Struthio camelus australis             NaN   

  2024 IUCN Red List category Synonyms Alternative common names  \
0                          LC      NaN                      NaN   
1                         NaN      NaN                      NaN   
2                         NaN      NaN                      NaN   
3                         NaN      NaN                      NaN   
4                         NaN      NaN                      NaN   

                                 Taxonomic source(s)    SISRecID  \
0  del Hoyo, J., Collar, N.J., Christie, D.A., El...  45020636.0   
1                                                NaN         NaN   
2                                                NaN         NaN   
3                                                NaN         NaN   
4                                                NaN         NaN   

   SpcRecID\n(used before SISRecID)    SubsppID  
0                         1016860.0         NaN  
1                               NaN  45020636_1  
2                               NaN  45020636_2  
3                               NaN  45020636_3  
4                               NaN  45020636_4

The data for birds is at subspecies level. It means that there is species name represented by column `Seq.`. All rows in same number represent one species.

2. Mammals

## Similarly, read file for mammals metadata.

In [14]:
loc = "/content/IUCN Mammals, Birds, Reptiles, and Amphibians.csv"
mammals_meta = pd.read_csv(loc, encoding='latin-1')
mammals_meta.head()

Species ID   Kingdom    Phylum     Class             Order       Family  \
0     42641.0  ANIMALIA  CHORDATA  MAMMALIA          RODENTIA      MURIDAE   
1  22687170.0  ANIMALIA  CHORDATA      AVES  CAPRIMULGIFORMES  TROCHILIDAE   
2     17879.0  ANIMALIA  CHORDATA  MAMMALIA          RODENTIA      MURIDAE   
3  47760825.0  ANIMALIA  CHORDATA  MAMMALIA          RODENTIA   CRICETIDAE   
4     42656.0  ANIMALIA  CHORDATA  MAMMALIA          RODENTIA  ABROCOMIDAE   

          Genus    Species                  Authority  Infraspecific rank  \
0     Abditomys   latidens            (Sanborn, 1952)                 NaN   
1      Abeillia   abeillei  (Lesson & DeLattre, 1839)                 NaN   
2   Abeomelomys      sevia    (Tate & Archbold, 1935)                 NaN   
3  Abrawayaomys    ruschii         Cunha & Cruz, 1979                 NaN   
4      Abrocoma  bennettii           Waterhouse, 1837                 NaN   

   ...                             Synonyms  \
0  ...                                  NaN   
1  ...                                  NaN   
2  ...  Melomys sevia|Pogonomelomys sevia     
3  ...                                  NaN   
4  ...                                  NaN   

                                  Common names (Eng)  Common names (Fre)  \
0                            Luzon Broad-toothed Rat                 NaN   
1                        Emerald-chinned Hummingbird                 NaN   
2  Menzies' Mouse, Menzies's Mouse, Highland Brus...                 NaN   
3                                       Ruschi's Rat                 NaN   
4                           Bennett's Chinchilla Rat                 NaN   

  Common names (Spa) Red List status Red List criteria  \
0                NaN              DD               NaN   
1                NaN              LC               NaN   
2                NaN              LC               NaN   
3                NaN              LC               NaN   
4                NaN              LC               NaN   

  Red List criteria version Year assessed Population trend  Petitioned  
0                       3.1        2008.0          unknown           N  
1                       3.1        2012.0       decreasing           N  
2                       3.1        2016.0          unknown           N  
3                       3.1        2016.0          unknown           N  
4                       3.1        2016.0          unknown           N  

[5 rows x 23 columns]

## For both birds and mammals answer the following questions.
1. Size of the datasets
2. For birds, filter the dataset using species-level rows only. Hint. Column `Subsp. Seq.` value of 0 indicates the main species name
3. How many bird species and mammal species are present in data frames?

In [49]:
#1-1
birds_meta.shape
# Birds_meta DataFrame has 33265 rows and 17 columns

(33265, 17)

In [50]:
#1-2
mammals_meta.shape
# Mammals_meta DataFrame has 16081 rows and 23 columns

(16081, 23)

In [51]:
#2
birds_species = birds_meta[birds_meta['Subsp. Seq.'] == 0]
birds_species.shape
# Species only has 11204 rows and 17 columns

(11204, 17)

In [52]:
#3-1
birds_species["Scientific name"].nunique()
# There are 11198 bird species present in the data frame

11198

In [53]:
#3-2
mammals_meta["Species"].nunique()
# There are 8779 mammal species present in the data frame

8779

# Database Integration Exercise
The main objective of this exercise is to create a comprehensive database by combining three data frames. Follow these steps  n your analysis.

## Create a concept map:  
* Illustrate the relationships between the three data frames.  
* Identify and highlight the Primary Keys in each data frame.  
* Use a diagramming tool (e.g., PowerPoint, LucidChart, or Draw.io) to create a visual representation of these connections.  
* Pay special attention to how you plan to merge the Mammals and Birds data frames.  

## Perform a vertical join of Mammals and Birds data frames:  
* Identify and select only the common columns present in both data frames.  
* Use `pd.concat()` or `df.append()` to create a larger dataframe called **metadata**.  
* Ensure the resulting dataframe has the correct number of rows and columns.  

In [82]:
birds_species = birds_species.copy()
mammals_meta = mammals_meta.copy()

birds_species["Scientific Name"] = birds_species["Scientific Name"]
mammals_meta["Scientific Name"] = mammals_meta["Species"]

In [87]:
birds_common = birds_species[["Scientific Name", "Family", "Order"]].copy()
mammals_common = mammals_meta[["Scientific Name", "Family", "Order"]].copy()

In [88]:
metadata = pd.concat([birds_common, mammals_common], axis=0)
metadata.shape

(27285, 3)

In [81]:
metadata.head()

Scientific Name     Family             Order Species
0         Struthio camelus  Ostriches  STRUTHIONIFORMES     NaN
5   Struthio molybdophanes  Ostriches  STRUTHIONIFORMES     NaN
6           Rhea americana      Rheas  STRUTHIONIFORMES     NaN
12            Rhea pennata      Rheas  STRUTHIONIFORMES     NaN
16      Nothocercus julius   Tinamous  STRUTHIONIFORMES     NaN

In [89]:
metadata.columns.tolist()

['Scientific Name', 'Family', 'Order']

In [92]:
birds_species.columns.tolist()
# There are 2 "Scientific Name" with one having the n on name being uppercase and one with n being lowercase, so they will be combined)

['Seq.',
 'Subsp. Seq.',
 'Order',
 'Family name',
 'Family',
 'Subfamily',
 'Tribe',
 'Common name',
 'Scientific name',
 'Authority',
 '2024 IUCN Red List category',
 'Synonyms',
 'Alternative common names',
 'Taxonomic source(s)',
 'SISRecID',
 'SpcRecID\n(used before SISRecID)',
 'SubsppID',
 'Scientific Name']

In [93]:
mammals_meta.columns.tolist()

['Species ID',
 'Kingdom',
 'Phylum',
 'Class',
 'Order',
 'Family',
 'Genus',
 'Species',
 'Authority',
 'Infraspecific rank',
 'Infraspecific name',
 'Infraspecific authority',
 'Stock/subpopulation',
 'Synonyms',
 'Common names (Eng)',
 'Common names (Fre)',
 'Common names (Spa)',
 'Red List status',
 'Red List criteria',
 'Red List criteria version',
 'Year assessed',
 'Population trend',
 'Petitioned',
 'Scientific Name']

## Merge the metadata dataframe with the Flavivirus dataframe:  
* Experiment with different join types: `inner`, `outer`, `left`, and `right`.  
* For each join type, record the resulting dataframe's shape (rows and columns).  
* Determine the **theoretically correct size of the merged dataframe if done manually**.  
* Compare the results of each join type with the expected outcome.  
* Choose the most appropriate join type based on your analysis.  

In [90]:
#inner
inner_merge = pd.merge(df, metadata, on='Scientific Name', how='inner')
inner_merge.shape
# There are 307 rows and 12 columns (2 were "Scientific Name" with one having the n on name bring uppercase and one with lowercase, so they are combined now)

(307, 12)

In [91]:
#outer
outer_merge = pd.merge(df, metadata, on='Scientific Name', how='outer')
outer_merge.shape
# There are 27616 rows and 12 columns

(27616, 12)

In [94]:
#left
left_merge = pd.merge(df, metadata, on='Scientific Name', how='left')
left_merge.shape
# There are 570 rows and 12 columns

(570, 12)

In [95]:
#right
right_merge = pd.merge(df, metadata, on='Scientific Name', how='right')
right_merge.shape
# There are 27353 rows and 12 columns

(27353, 12)

# Since the Flavivirus dataframe has 570 rows, and the metadata dataframe has 27285 rows, a ***left join*** preserves all 570 rows from df and adds metadata where available, which matches the intended purpose of merging supplementary data into the core dataset. An innter join only returned 307 rows, which is smaller than the original df, and the right join and outer join include extra rows from metadata that are not present in df, so none of those are appropriate. Therefore, ***LEFT JOIN*** is the correct merge.

### Create a pivote table,
1. Rows are virueses  
2. Columns are Taxonomic orders  
3. Values are number of unique host species  

In [101]:
pivot = df_pivot.pivot_table(index='Virus', columns='Order', values='Scientific Name', aggfunc=pd.Series.nunique)
pivot

Order                             ACCIPITRIFORMES  ANSERIFORMES  \
Virus                                                             
Bagaza virus                                  NaN           NaN   
Ilheus virus                                  NaN           NaN   
Japanese encephalitis virus                   NaN           NaN   
Koutango virus                                NaN           NaN   
Kunjin virus                                  NaN           NaN   
Louping-ill virus                             NaN           NaN   
Murray Valley encephalitis virus              NaN           NaN   
Rocio virus                                   NaN           NaN   
St. Louis Encephalitis virus                  1.0           NaN   
Tembusu virus                                 NaN           1.0   
Tickborne encephalitis virus                  1.0           7.0   
Uganda S virus                                NaN           NaN   
Usutu virus                                   4.0           NaN   
West Nile virus                              16.0           4.0   

Order                             BUCEROTIFORMES  CAPRIMULGIFORMES  \
Virus                                                                
Bagaza virus                                 NaN               NaN   
Ilheus virus                                 NaN               NaN   
Japanese encephalitis virus                  NaN               NaN   
Koutango virus                               NaN               NaN   
Kunjin virus                                 NaN               NaN   
Louping-ill virus                            NaN               NaN   
Murray Valley encephalitis virus             NaN               NaN   
Rocio virus                                  NaN               NaN   
St. Louis Encephalitis virus                 NaN               NaN   
Tembusu virus                                NaN               NaN   
Tickborne encephalitis virus                 NaN               NaN   
Uganda S virus                               NaN               NaN   
Usutu virus                                  1.0               1.0   
West Nile virus                              1.0               NaN   

Order                             CATHARTIFORMES  CHARADRIIFORMES  \
Virus                                                               
Bagaza virus                                 NaN              NaN   
Ilheus virus                                 NaN              NaN   
Japanese encephalitis virus                  NaN              NaN   
Koutango virus                               NaN              NaN   
Kunjin virus                                 NaN              NaN   
Louping-ill virus                            NaN              NaN   
Murray Valley encephalitis virus             NaN              NaN   
Rocio virus                                  NaN              NaN   
St. Louis Encephalitis virus                 NaN              1.0   
Tembusu virus                                NaN              NaN   
Tickborne encephalitis virus                 NaN              2.0   
Uganda S virus                               NaN              NaN   
Usutu virus                                  NaN              3.0   
West Nile virus                              2.0              9.0   

Order                             CICONIIFORMES  COLUMBIFORMES  CORACIIFORMES  \
Virus                                                                           
Bagaza virus                                NaN            NaN            NaN   
Ilheus virus                                NaN            NaN            NaN   
Japanese encephalitis virus                 NaN            2.0            NaN   
Koutango virus                              NaN            NaN            NaN   
Kunjin virus                                NaN            NaN            NaN   
Louping-ill virus                           NaN            NaN            NaN   
Murray Valley encephalitis virus            NaN      